language/vm/src/check_bounds.rs

```diff
@@ -121,8 +165,8 @@ impl<'a> BoundsChecker<'a> {
     }
 }
 
-pub trait BoundsCheck {
-    fn check_bounds(&self, module: &CompiledModuleMut) -> Vec<VMStaticViolation>;
+pub trait BoundsCheck<Context: Copy> {
```

Comment:
I think this trait is starting to become a bit unwieldy and induce a lot of boilerplate code. Instead of doubling down on the `BoundsCheck` abstraction (which doesn't seem to be buying us much), I think it might make more sense to eliminate it instead.

It looks like the trait is only used in `verify_impl`, and we could get rid of it by restructuring `verify` a bit.

Reply:
This trait does help make the code a little bit nicer when dealing with `SignatureToken` and `LocalsSignature` -- I can have a little bit of "overloading" for `SignatureToken`, and a generic wrapper that handles all the versions automatically for `LocalsSignature`.

But I agree with you that this trait introduces boilerplate code for things that have a fixed context (functions, structs etc.) While the ability to implement bounds checking in a potentially different environment sounds nice, it doesn't seem like we're expecting that to happen soon.

Anyway I'm fine with both.

Reply:
This PR is getting plenty of love.  I will stay out of it.  But I am curious to see what the code looks like once the BoundsCheck trait gets eliminated.  I also wish I understood when traits are good and when they are unwieldy.  Perhaps a conversation for later.

Reply:
@sblackshear I had another look and think maybe we should just keep the trait and move on.

Fundamentally, I don't care, as both paradigms rank very similarly to me. Removing the trait will remove some boilerplate (trait declaration & impls) code, but introduce more at other places:
  - Longer function names like `check_signature_token_bounds`.
  - We'll need to expand `verify_impl`, repeating the `.iter().enumerate().map(...)` pattern.
  - The trait approach makes a slight difference allowing logic/boilerplate code to be done at trait declaration/impls, rather than at the call sites. 

Anyway I don't think it's buying us much switching from either side to the other.